### PR TITLE
Feat: Helm Chart Local Installation Support

### PR DESCRIPTION
**Pull request by daniel-hutao**
## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [x] I have added relevant tests

## Description

New feat with helm chart local installation

## Related Issues

#1084 

## New Behavior (screenshots if needed)

- Apply without Internet

image

- Config file was used

```yaml
---
# core config
varFile: "" # If not empty, use the specified external variables config file
toolFile: "" # If not empty, use the specified external tools config file
pluginDir: "~/.devstream/plugins"
state: # state config, backend can be local or s3
  backend: local
  options:
    stateFile: devstream.state

---
tools:
# name of the tool
- name: jenkins
  # id of the tool instance
  instanceID: default
  # format: name.instanceID; If specified, dtm will make sure the dependency is applied first before handling this tool.
  dependsOn: [ ]
  # options for the plugin
  options:
    # Helm repo information
    repo:
      # name of the Helm repo
      name: jenkins
      # url of the Helm repo
      url: https://charts.jenkins.io
    # Helm chart information
    chart:
      # local path of the chart; if chartPath != "", repo.name and repo.url will be ignored. e.g. "foo.tgz", "./foo.tgz", "/tmp/foo.tgz"
      chartPath: "./jenkins-4.2.1.tgz"
      # name of the chart
      chartName: jenkins/jenkins
      # k8s namespace where jenkins will be installed
      namespace: jenkins
      # whether to wait for the release to be deployed or not
      wait: true
      # the time to wait for any individual Kubernetes operation (like Jobs for hooks). This defaults to 5m0s
      timeout: 5m
      # whether to perform a CRD upgrade during installation
      upgradeCRDs: true
      # custom configuration. You can refer to [Jenkins values.yaml](https://github.com/jenkinsci/helm-charts/blob/main/charts/jenkins/values.yaml)
      valuesYaml: |
        persistence:
          storageClass: ""
        serviceAccount:
          create: true
          name: jenkins
        controller:
          adminUser: "admin"
          adminPassword: "changeme"
          serviceType: NodePort
          nodePort: 32000
```

### CODE REVIEW

File: pkg/util/helm/local.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	sFile, err := os.Stat(chartPath)
+	if err != nil {
+		return fmt.Errorf("chart <%s> doesn't exist", chartPath)
+	}
+	if !sFile.Mode().IsRegular() {
+		return fmt.Errorf("the chart file <%s> is non-regular (%q)", chartPath, sFile.Mode().String())
+	}
+
+	// destination chart path checks
+	dFilePath := filepath.Join(repositoryCache, chartPath)
```
Comment (iyear) [inline pkg/util/helm/local.go:29]: Use `os.TempDir()` instead of hard-coding `/tmp`

Comment (iyear) [inline pkg/util/helm/local.go:29]: Use `os.TempDir()` instead of hard-coding `/tmp`

File: pkg/util/helm/local.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+			return err
+		}
+		if os.SameFile(sFile, dFile) {
+			log.Debugf("The source chart package and destination chart package are same.")
+			return nil
+		}
+		// check md5
+		equal, err := md5.FilesMD5Equal(chartPath, dFilePath)
+		if err != nil {
+			log.Errorf("Got error: %s.", err)
```
Comment (iyear) [inline pkg/util/helm/local.go:50]: Just wrap err and return it upstream to output the error.

File: pkg/util/helm/local.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	dFile, err := os.Stat(dFilePath)
+	if err != nil { // return err if err != nil and err != "NotExist"
+		if !os.IsNotExist(err) {
+			log.Errorf("Got error: %s.", err)
+			return err
+		}
+	} else { // err == nil -> file exists -> check if the source file equals destination file
+		if !(dFile.Mode().IsRegular()) {
+			err = fmt.Errorf("the destination file <%s> is non-regular (%q)", dFilePath, dFile.Mode().String())
+			log.Errorf("Got error: %s.", err)
```
Comment (iyear) [inline pkg/util/helm/local.go:40]: Just wrap err and return it upstream to output the error.

File: pkg/util/helm/local.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+			return err
+		}
+		if equal {
+			log.Infof("The chart package already exists in the cache directory.")
+			return nil
+		}
+		// remove the destination file if its name equals to source file but their contents don't equal
+		if err = os.RemoveAll(dFilePath); err != nil {
+			log.Errorf("Got error: %s.", err)
```
Comment (iyear) [inline pkg/util/helm/local.go:59]: Just wrap err and return it upstream to output the error.

File: pkg/util/helm/local.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	}()
+
+	// create destination file
+	dFile, err := os.Create(dstFile)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if closeErr := dFile.Close(); closeErr != nil {
+			log.Errorf("Failed to close file %s: %s", dstFile, closeErr)
```
Comment (iyear) [inline pkg/util/helm/local.go:94]: Just wrap err and return it upstream to output the error.

**Review by iyear — APPROVED** [review]
LGTM 👏

**Comment by daniel-hutao** [comment]
@iyear I've updated. Thanks for your great feedback.

File: internal/pkg/plugin/argocd/argocd.go [hunk 6]
```diff
@@ -8,6 +8,7 @@ import (
 
 var defaultHelmConfig = helm.Options{
 	Chart: helmCommon.Chart{
+		ChartPath:   "",
```
Comment (aFlyBird0) [inline internal/pkg/plugin/argocd/argocd.go:11]: `""` is zero value of `string`, so it is unnecessary to set the default value here.

But on the other hand, you can write in this way to emphasize it.

File: pkg/util/helm/local.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+	// destination chart path is empty, then create it.
+	log.Debugf("Prepare to copy <%s> to <%s>.", chartPath, dFilePath)
+	if err = os.MkdirAll(repositoryCache, 0755); err != nil {
+		return err
+	}
+	return copyFile(chartPath, dFilePath)
+}
+
+func copyFile(srcFile, dstFile string) (err error) {
```
Comment (aFlyBird0) [inline pkg/util/helm/local.go:67]: There is already a function named `CopyFile` in `/pkg/util/file/file.go`, you could combine them.

**Comment by aFlyBird0** [comment]
I‘m reviewing, please do not merge until my review is finished.

File: pkg/util/helm/local.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+	// destination chart path checks
+	dFilePath := filepath.Join(repositoryCache, chartPath)
+	log.Debugf("The destination chart path is <%s>.", dFilePath)
+	dFile, err := os.Stat(dFilePath)
+	if err != nil { // return err if err != nil and err != "NotExist"
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("stat file failed: %s.", err)
+		}
+	} else { // err == nil -> file exists -> check if the source file equals destination file
```
Comment (aFlyBird0) [inline pkg/util/helm/local.go:36]: nice comment 👏

**Review by aFlyBird0 — CHANGES_REQUESTED** [review]

File: internal/pkg/plugin/argocd/argocd.go (hunk 6 above)
Comment (daniel-hutao) [reply internal/pkg/plugin/argocd/argocd.go:11]: yes, emphasize only

File: pkg/util/helm/local.go (hunk 8 above)
Comment (daniel-hutao) [reply pkg/util/helm/local.go:36]: Thx

File: pkg/util/helm/local.go (hunk 7 above)
Comment (daniel-hutao) [reply pkg/util/helm/local.go:67]: I‘ll do it later

**Comment by daniel-hutao** [comment]
@aFlyBird0 Please take a look again. I've merge the copy function in c28ef1b0012e5cd0f251dd4e45d4d09fb47572dc

**Review by aFlyBird0 — APPROVED** [review]

**Pull request merged**